### PR TITLE
Added Template Runner

### PR DESCRIPTION
**Pull request by matthew-rollitt**
Small program that can consume a json file and create and tests jobs/pools on them. If the job fails the container should be left over

### CODE REVIEW

**Comment by Tatsinnit** [comment]
💡 📔   @matthew-rollitt : Please write context of this work in detail for the Reviewers to understand the file reference and their importance. And little gist for this work as well please.

* Before merging you can summaries it.

File: Runner/JobManager.py
Comment (Tatsinnit) [inline Runner/JobManager.py]: :bulb: **Suggestion: Minor please take care moving forward with final version** Unit Tests? for this entire file will be great idea. All the public functs should hooked up to UTs

File: Runner/JobManager.py
Comment (Tatsinnit) [inline Runner/JobManager.py:2]: :bulb: **minor** Same import multiple times: see line : 17

File: Runner/JobManager.py
Comment (Tatsinnit) [inline Runner/JobManager.py:9]: 🗒 **minor** note:  you can decide what length will be good for this work, but a specific standard will help in longer run for this project. some of PEP 8 here: https://stackoverflow.com/questions/88942/why-does-pep-8-specify-a-maximum-line-length-of-79-characters

File: Runner/Utils.py
Comment (Tatsinnit) [inline Runner/Utils.py]: :bulb: 🐛 : what is the `async` behavior this funct is trying to do? **apart** from `async.sleep` on line242` which seems like might be related to any scenario? also this returns the tuple or some sort. i.e. `True or False, success or value error.`

File: Runner/Utils.py
Comment (Tatsinnit) [inline Runner/Utils.py:219]: 📓 💡 Please note: this will return list of `incomplete_tasks` and then eventually **no operation** is happening on this list? and code returns with True or false, is this correct?

File: Runner/Utils.py
Comment (Tatsinnit) [inline Runner/Utils.py]: 💡 **Important** Please note: this is `async` method but actual operation within this method is all non async, i.e. what is awaitable in this method? [helpful link](https://docs.python.org/3/library/asyncio.html) if this funct will get yielded somehwere above this funct needs to have async behavior. _Please refer to a sample I created_.

File: Runner/Utils.py
Comment (Tatsinnit) [inline Runner/Utils.py:77]: 💡 **Minor** Is this intentional?

File: Runner/JobManager.py
Comment (Tatsinnit) [inline Runner/JobManager.py]: 🏗 **Something to think** What is the `returntype` of this?

File: Runner/Utils.py
Comment (Tatsinnit) [reply Runner/Utils.py]: For example: Please see your `RunJob` implementation which has await on `submit_job` which does has await on `batch_service_client.job.add, job`

File: Runner/Utils.py
Comment (Tatsinnit) [inline Runner/Utils.py]: 💡 **minor** This might help further: https://www.python.org/dev/peps/pep-0492/  also if you wish to differentiate with `private` method this link might come handy: https://stackoverflow.com/questions/17193457/private-methods-in-python

File: Runner/JobManager.py
Comment (Tatsinnit) [inline Runner/JobManager.py]: 💡 **Important**

**Sample for better understanding** this is just vanilla i created and you can play around with the idea. :)

image



```
import asyncio
async def get_programmer_name(name):
    await asyncio.sleep(3)
    return "programmer-%s" % name
async def foobar():
    result = await get_programmer_name("matt")
    print("---return {}r---".format(result))
    return result
def foomain():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    result = loop.run_until_complete(foobar())
    return result
programmer_name = foomain()
---return programmer-mattr---
```

**Review by Tatsinnit — CHANGES_REQUESTED** [review]

File: Runner/JobManager.py
Comment (Tatsinnit) [inline Runner/JobManager.py:36]: 💡 📓 **Minor recommendation**: Python now support `type annotations` so that can help in specifying type of input to the function and its return types. For longer run this will help in organical growth of this file, and Unit test writeups become easier: for example:

```
async def monty(
   self,
   obj_1: str
   obj_2: some_type
) -> return_type :
   funct_body
   pass
```

File: Runner/Runner.py
Comment (Tatsinnit) [inline Runner/Runner.py]: 💡 📓 **Future good practice** Some of these are unused imports, I would encourage to use `flake 8` its very imple to run on your file and it will catch few things for you. Flake 8 link : http://flake8.pycqa.org/en/latest/

File: Runner/Runner.py
Comment (Tatsinnit) [inline Runner/Runner.py:73]: 💡 **Important** do we need `try - exept` here? seems like this is assignment operation?

File: Runner/Runner.py
Comment (Tatsinnit) [inline Runner/Runner.py:1]: 💡 **Recommendation**: Unt tests? they will come very handy in longer run.

File: Runner/Utils.py
Comment (Tatsinnit) [inline Runner/Utils.py]: 💡**Important** Is this present only to make this fucntion `async`? or any contextual reason for `sleep(1)` here for this method?

File: Runner/Utils.py
Comment (Tatsinnit) [inline Runner/Utils.py:246]: 💡 **Important** In this context is `f.name` a list or just `string with finlename` if latter is the case then you can just do `==` instead of `in` use.

File: Runner/Runner.py
Comment (Tatsinnit) [inline Runner/Runner.py:76]: 💡 **Minor: just idea share**: you can either return some sort of `invalid` status here or some exit mechanism, using `pass` will jsut silently run the whole on on except `keyError` in this instance. There are few other places with similar pattern.

File: Runner/JobManager.py
Comment (matthew-rollitt) [reply Runner/JobManager.py]: But isn't this a testing program, If I do that wouldn't I be writing UTs for UTs?

File: Runner/JobManager.py
Comment (matthew-rollitt) [inline Runner/JobManager.py]: yep can do

File: Runner/JobManager.py
Comment (Tatsinnit) [reply Runner/JobManager.py]: 🐛 This seems not true, You have **No** `asserts` inside `JobRunner` ? if this is true then UTs will usually have some sort of `assert` of some sort. 

* Also if the intention is to make `jobrunner` a UT then you should follow a convetion of starting file as: `Test_` 

But please take that as a note: Reason behind UT is as follows:
* This will introduce better file convention in your project.
* Also JobRunner has no `asserts` so definetly this is some sort of runner program, having test makes it easy for you in longer run. :)

File: Runner/Runner.py
Comment (matthew-rollitt) [reply Runner/Runner.py]: Thanks, I'll use flake 8. I was wondering what the best tool is for the job.

File: Runner/JobManager.py
Comment (Tatsinnit) [reply Runner/JobManager.py]: :bulb: **Important** Please note: Sample below is from async to Blocking call: using `run_in_executor`:

image

```
def blocking_io(arg1="hulk", arg2="smash"):
    monty = "{}{}".format(arg1, arg2)
    print("----- {}".format(monty))
    return monty
async def async_foo():
    loop = asyncio.get_event_loop()
    result = await loop.run_in_executor(
        None, blocking_io, "IronMan")
    print('default thread pool', result)
    return result
def foo():
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    result = loop.run_until_complete(async_foo())
    print("returned result is - {}".format(result))
    return result
avengers = foo()
----- IronMansmash
default thread pool IronMansmash
returned result is - IronMansmash

```

File: Runner/Utils.py
Comment (matthew-rollitt) [reply Runner/Utils.py]: We don't need to make this async since we are waiting for the task to complete.

File: Runner/Utils.py
Comment (matthew-rollitt) [reply Runner/Utils.py:219]: Correct, we are using this code to check to see if all the tasks are in a completed state. We check the output of the task in 'check_task_output'

File: Runner/Runner.py
Comment (matthew-rollitt) [reply Runner/Runner.py:76]: I can just drop the KeyError in the try/catch. I'm wanting to just continue on because that json element can be empty by design.

File: Runner/JobManager.py
Comment (Tatsinnit) [reply Runner/JobManager.py]: 📔  -> follow up ping on this, is this sorted, I checked recent commit seems like not catering the exact use of async used between : `main call -> to Validate -> Wait` method. Play around and please let me know and we can sit down and make small changes.

💡 : Future reference only: Best for these scenarios moving forwards will be make small test files add UTs for your scenario so that your functs are contolled.

**Comment by Tatsinnit** [comment]
💡 Reference Note: We will not merge this, mainly intention was the first pass, so once main issue is sorted we can close this PR, :) setting comment here, in case anyone accidently merges.

File: Runner/Utils.py
Comment (matthew-rollitt) [reply Runner/Utils.py:77]: yep, from what I was seeing online the best way to handle it when a json value doesn't exist is just except the error and move on

File: Runner/JobManager.py
Comment (matthew-rollitt) [reply Runner/JobManager.py:36]: Thanks I'll keep that in mind

File: Runner/Utils.py
Comment (matthew-rollitt) [reply Runner/Utils.py:246]: need to use in since we to deal with the filepath in the file

File: Runner/Runner.py
Comment (matthew-rollitt) [reply Runner/Runner.py:1]: I was thinking about adding tests to this and yes I do like adding tests to this. But in this case, I feel the purpose of this script is really just to test out all the templates. I don't feel like it's worth adding

File: Runner/JobManager.py
Comment (matthew-rollitt) [reply Runner/JobManager.py:2]: removed unneeded imports

**Pull request closed without merging**
